### PR TITLE
Fix #5023 - Add support for editing bookmarks

### LINKED ISSUE #5023: Add support for editing bookmarks
Finally.

### PR DESCRIPTION
**Pull request by justindarc**

### CODE REVIEW

File: Client/Frontend/Library/BookmarkDetailPanel.swift [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+class BookmarkDetailPanelError: MaybeErrorType {
+    public var description = "Unable to save BookmarkNode."
+}
+
+class BookmarkDetailPanel: SiteTableViewController {
+    enum BookmarkDetailSection: Int {
+        case fields
+        case folder
+
+        static let count = 2
```
Comment (garvankeeley) [inline Client/Frontend/Library/BookmarkDetailPanel.swift]: enum has allValues.count built-in

File: Client/Frontend/Library/BookmarkDetailPanel.swift [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+            self.title = Strings.BookmarksEditBookmark
+        } else if let bookmarkFolder = bookmarkNode as? BookmarkFolder {
+            self.bookmarkItemOrFolderTitle = bookmarkFolder.title
+
+            self.title = Strings.BookmarksEditFolder
+        }
+    }
+
+    convenience init(profile: Profile, withNewBookmarkNodeType bookmarkNodeType: BookmarkNodeType, parentBookmarkFolder: BookmarkFolder) {
```
Comment (garvankeeley) [inline Client/Frontend/Library/BookmarkDetailPanel.swift]: Why is this init needed?

File: Client/Frontend/Library/BookmarkDetailPanel.swift [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .cancel, closure: { _ in
```
Comment (garvankeeley) [inline Client/Frontend/Library/BookmarkDetailPanel.swift]: nit: trailing closure syntax would be ok here

File: Client/Frontend/Library/BookmarkDetailPanel.swift [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        // Handle folder selection cells.
+        guard indexPath.section == BookmarkDetailSection.fields.rawValue else {
+            guard let cell = tableView.dequeueReusableCell(withIdentifier: BookmarkDetailFolderCellIdentifier, for: indexPath) as? OneLineTableViewCell else {
+                return super.tableView(tableView, cellForRowAt: indexPath)
+            }
+
+            cell.imageView?.image = UIImage(named: "bookmarkFolder")?.createScaled(CGSize(width: BookmarkDetailPanelUX.FolderIconSize, height: BookmarkDetailPanelUX.FolderIconSize))
```
Comment (garvankeeley) [inline Client/Frontend/Library/BookmarkDetailPanel.swift]: Maybe change definition of `BookmarkDetailPanelUX.FolderIconSize = CGSize(width: xx, height:xx)`, so then createScaled(BookmarkDetailPanelUX.FolderIconSize)

File: Client/Frontend/Library/BookmarkDetailPanel.swift [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+                for case let childFolder as BookmarkFolder in folder.children ?? [] {
+                    // Any "root" folders (i.e. "Mobile Bookmarks") should
+                    // have an indentation of 0.
+                    if childFolder.isRoot {
+                        addFolder(childFolder)
+                    }
+                    // Otherwise, all non-root folder should increase the
+                    // indentation by 1.
+                    else {
+                        addFolder(childFolder, indent: indent + 1)
```
Comment (garvankeeley) [inline Client/Frontend/Library/BookmarkDetailPanel.swift]: What is the max indent level? This should be capped, at which point it stops indenting (which is ok for those outliers that have deeply nested folders, they can still find a way to use this).

File: Client/Frontend/Widgets/PhotonActionSheet/PhotonActionSheet.swift [hunk 6]
```diff
@@ -54,9 +54,13 @@ class PhotonActionSheet: UIViewController, UITableViewDelegate, UITableViewDataS
         self.closeButton.setTitle(closeButtonTitle, for: .normal)
     }
 
-    init(title: String? = nil, actions: [[PhotonActionSheetItem]], closeButtonTitle: String = Strings.CloseButtonTitle, style presentationStyle: UIModalPresentationStyle) {
+    init(title: String? = nil, actions: [[PhotonActionSheetItem]], closeButtonTitle: String = Strings.CloseButtonTitle, style presentationStyle: UIModalPresentationStyle? = nil) {
```
Comment (garvankeeley) [inline Client/Frontend/Widgets/PhotonActionSheet/PhotonActionSheet.swift]: Make this non-optional again please. The `nil == .centered` I find confusing. Or make the default .centered (`, style presentationStyle: UIModalPresentationStyle = .centered) `if that is what was needed here.

File: Client/Frontend/Widgets/TextFieldTableViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+extension TextFieldTableViewCell: UITextFieldDelegate {
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        if let text = textField.text,
+            let textRange = Range(range, in: text) {
+            let updatedText = text.replacingCharacters(in: textRange, with: string)
+            delegate?.textFieldTableViewCell(self, didChangeText: updatedText)
+        }
+        return true
```
Comment (garvankeeley) [inline Client/Frontend/Widgets/TextFieldTableViewCell.swift]: If it returns true every time, should this be end editing?

Comment (justindarc) [reply Client/Frontend/Widgets/TextFieldTableViewCell.swift]: No, this is just a way to get updated values from the textfield as they are entered.

File: Client/Frontend/Library/BookmarkDetailPanel.swift (hunk 2 above)
Comment (justindarc) [reply Client/Frontend/Library/BookmarkDetailPanel.swift]: Needed for adding items.

File: Client/Frontend/Library/BookmarkDetailPanel.swift (hunk 3 above)
Comment (justindarc) [reply Client/Frontend/Library/BookmarkDetailPanel.swift]: Done.

File: Client/Frontend/Library/BookmarkDetailPanel.swift (hunk 4 above)
Comment (justindarc) [reply Client/Frontend/Library/BookmarkDetailPanel.swift]: Done.

File: Client/Frontend/Library/BookmarkDetailPanel.swift (hunk 5 above)
Comment (justindarc) [reply Client/Frontend/Library/BookmarkDetailPanel.swift]: Added a commit to programmatically determine an appropriate max indent level depending on screen width.

File: Client/Frontend/Widgets/PhotonActionSheet/PhotonActionSheet.swift (hunk 6 above)
Comment (justindarc) [reply Client/Frontend/Widgets/PhotonActionSheet/PhotonActionSheet.swift]: Talked about this offline. There's a difference between `UIModalPresentationStyle` and `PresentationStyle` and this is the easiest way to get the `PresentationStyle` set to `.centered` without breaking other callers.

File: Client/Frontend/Library/BookmarkDetailPanel.swift (hunk 1 above)
Comment (justindarc) [reply Client/Frontend/Library/BookmarkDetailPanel.swift]: Added `CaseIterable` to get this. Done.

**Pull request merged**